File: src/client/engine/brush.ts
// Brushes — paint a single stroke into a context whose origin is the symmetry
// center and which has already had the image transform (rotation / reflection)
// applied. Coordinates inside a stroke are normalized; we multiply by `half`
// (the shorter half-axis in px) to get centered pixels. Shared by scene.ts
// (live + committed) and export.ts (PNG / replay).
//
// `strokeSegments` below is the ONE path builder. Live canvas, committed canvas,
// PNG/WebP/OG, SVG and replay all get their geometry from it, so a smoothed
// stroke cannot curve on screen and stay a polyline in a download. Canvas turns
// each segment into moveTo/lineTo or moveTo/bezierCurveTo; SVG turns the same
// list into `L` or `C` commands.

import { smoothStroke } from "../../shared/smooth";
import { REFERENCE_HALF, pressureAlpha, type Stroke, type Pt } from "../../shared/vector";

type AnyCtx = CanvasRenderingContext2D | OffscreenCanvasRenderingContext2D;

/** Pressure 0..1 → width multiplier, floored so strokes never vanish. */
function widthFactor(pressure: number): number {
  return 0.35 + 0.65 * Math.max(0, Math.min(1, pressure));
}

/**
 * One segment of a stroke's rendered path: always from source point `i` to
 * source point `i + 1`, optionally curved.
 *
 * Structurally a superset of `Cubic` from ../../shared/smooth, so the smoothed
 * builder's output is used verbatim rather than copied field by field.
 */
export interface StrokeSegment {
  /** Index of the source point this segment starts at. */
  i: number;
  /** End point — always equals the source point at `i + 1`. */
  x: number;
  y: number;
  /** Cubic control points. Absent on a straight segment. */
  c1x?: number;
  c1y?: number;
  c2x?: number;
  c2y?: number;
}

/**
 * THE shared path builder. Normalized coordinates, pure geometry, no `half`.
 *
 * Segment boundaries are the source points either way, because width and
 * spectrum hue already vary per source segment and must keep varying
 * identically — smoothing changes the shape between two points, never how many
 * there are.
 *
 * A stroke is smoothed only when it asks (`sm === 1`) AND has an interior to
 * smooth. `smoothStroke` returns null below 3 points, and that falls through to
 * the straight case: the coordinates are then copied from `pts` verbatim, with
 * no arithmetic, so a v1 stroke produces bit-identical output to the polyline
 * loop this replaced. That is what keeps every stored gallery piece rendering as
 * its saved PNG.
 */
export function strokeSegments(stroke: Stroke): StrokeSegment[] {
  const pts = stroke.pts;
  if (pts.length < 2) return [];
  if (stroke.sm === 1) {
    const cubics = smoothStroke(pts);
    if (cubics) return cubics;
  }
  const out: StrokeSegment[] = new Array(pts.length - 1);
  for (let k = 0; k < pts.length - 1; k++) {
    out[k] = { i: k, x: pts[k + 1][0], y: pts[k + 1][1] };
  }
  return out;
}

/**
 * The alpha a `po` stroke paints at, at a given pressure.
 *
 * `applyBrushMode` has already folded glow's ×0.7 into the base alpha, so
 * feeding that base through `pressureAlpha` is what makes the two factors
 * compose — glow at full pressure is still 0.7, not 1.
 */
function poAlpha(stroke: Stroke, pressure: number): number {
  const base = stroke.tool === "glow" ? stroke.opacity * 0.7 : stroke.opacity;
  return pressureAlpha(base, pressure);
}

/** Mean pressure over a stroke's points. Used where one alpha must stand in for
 *  the whole stroke — SVG, which cannot vary stroke-opacity along a path. */
export function meanPressure(pts: readonly Pt[]): number {
  if (pts.length === 0) return 1;
  let sum = 0;
  for (const p of pts) sum += p[2];
  return sum / pts.length;
}

/** Spectrum hue (deg) for a normalized point: hue follows the angle around the
 *  center, so a radial scribble rainbows automatically. */
export function spectrumHue(nx: number, ny: number): number {
  const a = Math.atan2(ny, nx); // -π..π
  return ((a / (Math.PI * 2)) * 360 + 360) % 360;
}

export function spectrumColor(nx: number, ny: number, opacity = 1): string {
  return `hsla(${spectrumHue(nx, ny).toFixed(1)}, 85%, 60%, ${opacity})`;
}

function applyBrushMode(ctx: AnyCtx, stroke: Stroke): void {
  ctx.lineCap = "round";
  ctx.lineJoin = "round";
  if (stroke.tool === "glow") {
    ctx.globalCompositeOperation = "lighter";
    ctx.globalAlpha = stroke.opacity * 0.7;
  } else {
    ctx.globalCompositeOperation = "source-over";
    ctx.globalAlpha = stroke.opacity;
  }
}

/**
 * Draw one stroke. The context must already be translated to center and have the
 * symmetry image transform applied. `half` is the shorter half-axis in px.
 * A single point renders as a dot.
 */
export function drawStroke(ctx: AnyCtx, stroke: Stroke, half: number): void {
  const pts = stroke.pts;
  if (pts.length === 0) return;

  const scale = half / REFERENCE_HALF;
  ctx.save();
  applyBrushMode(ctx, stroke);

  const isSpectrum = stroke.color === "spectrum";
  if (!isSpectrum) ctx.strokeStyle = stroke.color;

  if (pts.length === 1) {
    // dot
    const [nx, ny, p] = pts[0];
    const x = nx * half;
    const y = ny * half;
    const r = (stroke.size * scale * widthFactor(p)) / 2;
    ctx.beginPath();
    ctx.fillStyle = isSpectrum ? spectrumColor(nx, ny, 1) : stroke.color;
    // Without `po` this is the same assignment it always was — deliberately
    // written as a branch rather than folded into one expression, because the
    // v1 value is pinned by test/unit/render-trace.test.ts, which records every
    // globalAlpha write.
    ctx.globalAlpha =
      stroke.po === 1
        ? poAlpha(stroke, p)
        : stroke.tool === "glow"
          ? stroke.opacity * 0.7
          : stroke.opacity;
    ctx.arc(x, y, Math.max(0.5, r), 0, Math.PI * 2);
    ctx.fill();
    ctx.restore();
    return;
  }

  // Segment-by-segment so width (and spectrum hue) can vary along the stroke.
  // Each segment is stroked as its own path — chaining a smoothed stroke into a
  // single path would force one width and one hue on the whole thing.
  for (const seg of strokeSegments(stroke)) {
    const a = pts[seg.i];
    const b = pts[seg.i + 1];
    const w = stroke.size * scale * widthFactor((a[2] + b[2]) / 2);
    ctx.lineWidth = Math.max(0.5, w);
    if (isSpectrum) {
      const mx = (a[0] + b[0]) / 2;
      const my = (a[1] + b[1]) / 2;
      ctx.strokeStyle = spectrumColor(mx, my, 1);
    }
    // Guarded: an unconditional per-segment globalAlpha write would appear in
    // the v1 render trace, which is frozen. `po` is never set on a v1 stroke.
    if (stroke.po === 1) ctx.globalAlpha = poAlpha(stroke, (a[2] + b[2]) / 2);
    ctx.beginPath();
    ctx.moveTo(a[0] * half, a[1] * half);
    if (seg.c1x === undefined) {
      ctx.lineTo(seg.x * half, seg.y * half);
    } else {
      ctx.bezierCurveTo(
        seg.c1x * half,
        seg.c1y! * half,
        seg.c2x! * half,
        seg.c2y! * half,
        seg.x * half,
        seg.y * half,
      );
    }
    ctx.stroke();
  }

  ctx.restore();
}

/** Representative solid color for a stroke (used by the SVG exporter, which
 *  cannot vary color along a single <path>). */
export function representativeColor(stroke: Stroke): string {
  if (stroke.color !== "spectrum") return stroke.color;
  const pts: Pt[] = stroke.pts;
  if (pts.length === 0) return "#888888";
  // average position → one hue
  let sx = 0;
  let sy = 0;
  for (const [x, y] of pts) {
    sx += x;
    sy += y;
  }
  const h = spectrumHue(sx / pts.length, sy / pts.length);
  return hslToHex(h, 85, 60);
}

function hslToHex(h: number, s: number, l: number): string {
  s /= 100;
  l /= 100;
  const k = (n: number) => (n + h / 30) % 12;
  const a = s * Math.min(l, 1 - l);
  const f = (n: number) => {
    const color = l - a * Math.max(-1, Math.min(k(n) - 3, Math.min(9 - k(n), 1)));
    return Math.round(255 * color)
      .toString(16)
      .padStart(2, "0");
  };
  return `#${f(0)}${f(8)}${f(4)}`;
}

File: ios/KaleidoEngine/Sources/KaleidoEngine/Brush.swift
// Brush math — the pure, renderer-agnostic parts of src/client/engine/brush.ts.
// The actual painting (Core Graphics) lives in the app; these helpers keep the
// color/width math identical to the web so spectrum strokes and pressure scaling
// look the same on both platforms.

import Foundation

/// Pressure 0…1 → width multiplier, floored so strokes never vanish.
public func widthFactor(_ pressure: Double) -> Double {
    0.35 + 0.65 * max(0, min(1, pressure))
}

/// Spectrum hue (degrees, 0…360) for a normalized point — hue follows the angle
/// around the center, so a radial scribble rainbows automatically.
public func spectrumHue(nx: Double, ny: Double) -> Double {
    let a = atan2(ny, nx) // -π…π
    return ((a / (Double.pi * 2)) * 360 + 360).truncatingRemainder(dividingBy: 360)
}

/// Linear RGB-ish components in 0…1 (sRGB channel values, as CSS hsl() yields).
public struct RGB: Equatable, Sendable {
    public var r: Double
    public var g: Double
    public var b: Double
    public init(r: Double, g: Double, b: Double) {
        self.r = r
        self.g = g
        self.b = b
    }
}

/// HSL → RGB using the same algorithm as brush.ts `hslToHex` (h in degrees,
/// s and l in 0…100), so spectrum colors match the web exactly.
public func hslToRGB(h: Double, s: Double, l: Double) -> RGB {
    let sN = s / 100
    let lN = l / 100
    let a = sN * min(lN, 1 - lN)
    func f(_ n: Double) -> Double {
        let k = (n + h / 30).truncatingRemainder(dividingBy: 12)
        return lN - a * max(-1, min(k - 3, min(9 - k, 1)))
    }
    return RGB(r: f(0), g: f(8), b: f(4))
}

/// The spectrum color for a normalized point: fixed saturation 85%, lightness
/// 60%, matching the web's `hsla(hue, 85%, 60%)`.
public func spectrumRGB(nx: Double, ny: Double) -> RGB {
    hslToRGB(h: spectrumHue(nx: nx, ny: ny), s: 85, l: 60)
}

/// The single representative hue for a spectrum stroke (average point position),
/// used where a stroke needs one flat color. Mirrors `representativeColor`.
public func representativeSpectrumRGB(points: [StrokePoint]) -> RGB {
    guard !points.isEmpty else { return RGB(r: 0.53, g: 0.53, b: 0.53) } // #888888
    var sx = 0.0
    var sy = 0.0
    for p in points {
        sx += p.x
        sy += p.y
    }
    return hslToRGB(h: spectrumHue(nx: sx / Double(points.count), ny: sy / Double(points.count)), s: 85, l: 60)
}

/// Canvas background colors (hex), matching export.ts / scene.ts THEME.
public extension Background {
    var hex: String {
        switch self {
        case .light: return "#EEF0EC"
        case .dark: return "#13202A"
        }
    }
}

// ---- v2: pressure-scaled alpha -------------------------------------------

/// Alpha for a stroke carrying the `po` flag, at a given pressure.
///
/// The 0.25 floor is deliberate: a stroke whose lightest touch went fully
/// transparent would break up into disconnected blobs, which reads as a rendering
/// fault rather than as pressure. Strokes WITHOUT `po` never call this — they use
/// `opacity` flat, which is v1 behavior and is what keeps every existing gallery
/// piece pixel-identical.
public func pressureAlpha(opacity: Double, pressure: Double) -> Double {
    opacity * (0.25 + 0.75 * max(0, min(1, pressure)))
}

// ---- v2: stroke smoothing (mirrors src/shared/smooth.ts) -----------------

// Centripetal Catmull-Rom (α = 0.5) converted to cubic Béziers.
//
// Why centripetal rather than uniform (α = 0) or chordal (α = 1): pointer
// sampling is uneven — a fast flick leaves points far apart, a slow curl leaves
// them bunched. Uniform Catmull-Rom reacts to that unevenness by overshooting
// and, where two points nearly coincide, forming a visible cusp or a
// self-intersecting loop. Centripetal parameterization is the variant proven to
// produce neither for any input spacing, which is what makes it safe to apply to
// strokes we did not sample ourselves.
//
// The output is deliberately per-segment: one cubic from pts[i] to pts[i+1],
// tagged with i. Width and spectrum hue already vary per segment in the polyline
// renderer and must keep varying identically, so the smoothed path preserves the
// same segment boundaries rather than collapsing the stroke into one long path.

/// A cubic Bézier from source point `i` to source point `i + 1`.
public struct Cubic: Equatable, Sendable {
    /// Index of the source point this segment starts at.
    public let i: Int
    /// First control point.
    public let c1x: Double
    public let c1y: Double
    /// Second control point.
    public let c2x: Double
    public let c2y: Double
    /// End point — equals the source point at `i + 1`.
    public let x: Double
    public let y: Double

    public init(i: Int, c1x: Double, c1y: Double, c2x: Double, c2y: Double, x: Double, y: Double) {
        self.i = i
        self.c1x = c1x
        self.c1y = c1y
        self.c2x = c2x
        self.c2y = c2y
        self.x = x
        self.y = y
    }
}

/// Centripetal exponent. α = 0.5 is the whole point; see above.
public let SMOOTH_ALPHA: Double = 0.5

// Knot spans below this are treated as a coincident point. Coordinates are
// normalized to ~[-1, 1] and capture already drops moves under ~1.1px, so this
// only fires on hand-authored or pathological input — where its only job is to
// keep the arithmetic finite instead of producing NaN, which renders as nothing
// at all.
private let SMOOTH_EPS = 1e-9

/// Centripetal knot span between two points: |P1 - P0|^α, with α = 0.5.
private func knotSpan(_ ax: Double, _ ay: Double, _ bx: Double, _ by: Double) -> Double {
    let d = sqrt(hypot(bx - ax, by - ay))
    return d < SMOOTH_EPS ? SMOOTH_EPS : d
}

/// Build the smoothed path for a stroke's points.
///
/// Returns one cubic per source segment, in order. Fewer than 3 points has no
/// interior to smooth, so it returns nil and the caller draws the polyline it
/// would have drawn anyway — matching the rule that v1 strokes, which never carry
/// `sm`, keep rendering as polylines forever.
///
/// Tangents are computed once per POINT, as the knot-weighted average of the two
/// adjoining chord velocities, then scaled into each segment by that segment's
/// own knot span. Neighbouring segments therefore share a tangent direction while
/// having their own magnitude: the curve is G1 (visually smooth at every join)
/// rather than C1. Asserting C1 here would be asserting something false.
///
/// The first and last points have only one neighbour, so they take the one-sided
/// chord velocity. NOTE: PLAN §2.2 specified "endpoints duplicated" instead.
/// Duplicating a point makes its knot span zero, which turns the interior tangent
/// formula into 0/0; clamping that with an epsilon yields a tangent of
/// approximately-but-not-exactly zero, so the first control point lands on
/// -0.6199999998797687 where it should be -0.62. That residue is both a flat
/// start to every stroke and a cross-platform parity hazard — precisely the kind
/// of near-miss this port exists to rule out. The web took the one-sided chord
/// (T01, flagged for M1) and this follows it, because the two implementations
/// agreeing is the requirement.
public func smoothStroke(_ pts: [StrokePoint]) -> [Cubic]? {
    let n = pts.count
    if n < 3 { return nil }

    // Knot span for each segment k → k+1.
    var span = [Double](repeating: 0, count: n - 1)
    for k in 0..<(n - 1) {
        span[k] = knotSpan(pts[k].x, pts[k].y, pts[k + 1].x, pts[k + 1].y)
    }

    // Velocity (tangent per unit knot) at each point.
    var vx = [Double](repeating: 0, count: n)
    var vy = [Double](repeating: 0, count: n)

    vx[0] = (pts[1].x - pts[0].x) / span[0]
    vy[0] = (pts[1].y - pts[0].y) / span[0]
    vx[n - 1] = (pts[n - 1].x - pts[n - 2].x) / span[n - 2]
    vy[n - 1] = (pts[n - 1].y - pts[n - 2].y) / span[n - 2]

    for k in 1..<(n - 1) {
        let dPrev = span[k - 1]
        let dNext = span[k]
        // Chord velocities either side of the point.
        let inX = (pts[k].x - pts[k - 1].x) / dPrev
        let inY = (pts[k].y - pts[k - 1].y) / dPrev
        let outX = (pts[k + 1].x - pts[k].x) / dNext
        let outY = (pts[k + 1].y - pts[k].y) / dNext
        // Weighted toward the SHORTER side, which is what keeps a tight corner
        // from being smoothed into a bulge that leaves the points' bounding box.
        let total = dPrev + dNext
        vx[k] = (inX * dNext + outX * dPrev) / total
        vy[k] = (inY * dNext + outY * dPrev) / total
    }

    var out: [Cubic] = []
    out.reserveCapacity(n - 1)
    for k in 0..<(n - 1) {
        let d = span[k] / 3
        out.append(Cubic(
            i: k,
            c1x: pts[k].x + vx[k] * d,
            c1y: pts[k].y + vy[k] * d,
            c2x: pts[k + 1].x - vx[k + 1] * d,
            c2y: pts[k + 1].y - vy[k + 1] * d,
            x: pts[k + 1].x,
            y: pts[k + 1].y
        ))
    }
    return out
}
